File: Lab/Lab6/task2.m
clear all;close all;clc;

N = 12;
fc = 1000;
fs = 4*fc;
mu = 0;sigma = sqrt(0.01);
duration_signal = 29;amp_carrier = 2;
BW_lpf = 1.5*N;BW_chanl = 0.5*N;

for T = 0:duration_signal
    t = -0.5:1/fs:0.5;
    
    m1t = cos(2*pi*N*t);
    m2t = 2*N*sinc(2*N*t);
    m3t = 200*(cos(200*pi*t)./(1-40000*t.*t)).*sinc(200*t);
    noise = mu + sigma*randn(1,length(t));
    chanl = 2*BW_chanl*sinc(2*BW_chanl*t).*(2*cos(2*pi*fc*t));
    carrier = amp_carrier*cos(2*pi*fc*t);
    
    signals = {m1t;m2t;m3t}; take = randi(3);
    m_t = cell2mat(signals(take));
    modulated = (m_t + 44).*carrier;
    modulated = modulated + noise;
    modulated = conv(modulated,chanl,'same');
    demodulated = hilbert(modulated).*exp(-1i*2*pi*fc*t);
    
    % Modulated signal plot
    nfft = length(modulated);m_f = fft(modulated,nfft)/fs;
    freqaxis = linspace(-fs/2, fs/2, nfft);
    
    figure(1);hold all ;
    subplot(2,1,1), plot(t+0.5+T,modulated);
    title('Modulated Signal');xlabel('Time(sec)');
    ylabel('Amplitude(volts)');grid on;hold on;    
    subplot(2,1,2), plot(freqaxis,fftshift(abs(m_f)));
    xlabel('Freq(Hz)');ylabel('Magnitude');grid on;
    
    % Demodulated signal plot
    nfft = length(demodulated);
    m_f = fft(demodulated,nfft)/fs;
    freqaxis = linspace(-fs/2, fs/2, nfft);
  
    figure(2);hold all ;subplot(2,1,1), plot(t+0.5+T,demodulated);
    title('Demodulated signal');xlabel('Time(sec)');ylabel('Amplitude(volts)');
    grid on;hold on;subplot(2,1,2), plot(freqaxis,fftshift(abs(m_f)));
    xlabel('Freq(Hz)');ylabel('Magnitude');grid on;
    pause(0.5);
end

File: Lab/Lab7/task2.py
import numpy as np
import matplotlib.pyplot as plt
import random
from scipy.signal import hilbert

signal_duration=10

for T in range(10):
 t_start=-0.5
 t_stop=0.5
 carrier_freq=500
 fs=10*carrier_freq
 ts=1/fs
 time=np.arange(t_start,t_stop,ts)
 A=25
 carrier_t=A*np.cos(2*np.pi*carrier_freq*time)
 carrier_f=np.fft.fftshift(abs(np.fft.fft(carrier_t)/fs))
 len_time=len(time)
 freq_axis=np.linspace(-fs/2,fs/2,len_time)
 B=random.randint(1,5)
 m_t=20*B*np.sinc(20*B*time)
 m_f=np.fft.fftshift(abs(np.fft.fft(m_t)/fs))
 mod_t=(1+m_t/A)*carrier_t
 mod_f=np.fft.fftshift(abs(np.fft.fft(mod_t)/fs))
 mu=0
 sigma_sq=0.001
 sigma=np.sqrt(sigma_sq)
 n_t=mu+sigma*np.random.randn(len(time))
 op_t=0.01*mod_t+n_t
 op_f=np.fft.fftshift(abs(np.fft.fft(op_t)/fs))
 op_demod_t=abs(hilbert(op_t))-A*0.01
 op_demod_f=np.fft.fftshift(abs(np.fft.fft(op_demod_t)/fs))

 plt.figure(1)
 plt.subplot(5,1,1)
 plt.plot(time+T,m_t)
 plt.title('Message signal-time domain')
 plt.xlabel('Time(in sec)')
 plt.ylabel('Amplitude')

 plt.figure(1)
 plt.subplot(5, 1, 2)
 plt.plot(time + T, carrier_t)
 plt.title('Carrier Signal-time domain')
 plt.xlabel('Time(in sec)')
 plt.ylabel('Amplitude')

 plt.figure(1)
 plt.subplot(5, 1, 3)
 plt.plot(time + T, mod_t)
 plt.title('Modulated signal-time domain')
 plt.xlabel('time(in sec)')
 plt.ylabel('Amplitude')

 plt.figure(1)
 plt.subplot(5, 1, 4)
 plt.plot(T+time, op_t)
 plt.title('Modulated signal after channel time domain')
 plt.xlabel('time(in sec)')
 plt.ylabel('Amplitude')

 plt.figure(1)
 plt.subplot(5, 1, 5)
 plt.plot(time + T, op_demod_t)
 plt.title('Demodulated signal-time domain')
 plt.xlabel('time(t)')
 plt.ylabel('Amplitude')

 plt.figure(2)
 plt.subplot(5, 1, 1)
 plt.plot(freq_axis, m_f)
 plt.title('Message signal-frequency domain')
 plt.xlabel('Frequency (in Hz)')
 plt.ylabel('Magnitude')

 plt.figure(2)
 plt.subplot(5, 1, 2)
 plt.plot(freq_axis, carrier_f)
 plt.title('Carrier signal-frequency domain')
 plt.xlabel('Frequency(in Hz)')
 plt.ylabel('Magnitude')

 plt.figure(2)
 plt.subplot(5, 1, 3)
 plt.plot(freq_axis, mod_f)
 plt.title('Modulated signal-frequency domain')
 plt.xlabel('Frequency(in Hz)')
 plt.ylabel('Magnitude')

 plt.figure(2)
 plt.subplot(5, 1, 4)
 plt.plot(freq_axis, op_f)
 plt.title('Modulated signal after channel frequency domain')
 plt.xlabel('Frequency(in Hz)')
 plt.ylabel('Magnitude')

 plt.figure(2)
 plt.subplot(5, 1, 5)
 plt.plot(freq_axis, op_demod_f)
 plt.title('Demodulated signal frequency domain')
 plt.xlabel('Frequency(in Hz)')
 plt.ylabel('Magnitude')
 plt.pause(0.5)

plt.show()